Handle failed IOSurface creation for image buffer

nmoucht opened the pull request:
#### e87a1dc9e5b517e7aab9aea3163dc787943b08de
<pre>
Handle failed IOSurface creation for image buffer
<a href="https://bugs.webkit.org/show_bug.cgi?id=242556">https://bugs.webkit.org/show_bug.cgi?id=242556</a>
&lt;rdar://94622888&gt;

Reviewed by Said Abou-Hallawa.

Check the backend handle sent back to the web process to ensure the image buffer in the web process
is in synch with the GPU process. For the test, create a canvas greater than the max IOSurface size,
disabling the check in the web process so that an accelerated buffer is created, IOSurface
creation in the GPU process fails so it creates an unaccelerated buffer, and ensure that the
web process correctly changes the backend type to an unaccelerated buffer.

* LayoutTests/fast/canvas/image-buffer-iosurface-disabled-expected.txt: Added.
* LayoutTests/fast/canvas/image-buffer-iosurface-disabled.html: Added.
* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::shouldBlockIOSurfaceCreationForNextImageBuffer):
* Source/WebCore/platform/graphics/ConcreteImageBuffer.h:
* Source/WebCore/platform/graphics/ImageBuffer.h:
(WebCore::ImageBuffer::fallbackToUnaccelerated):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::shouldBlockIOSurfaceCreationForNextImageBuffer):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/GPUProcess/graphics/RemoteImageBuffer.h:
(WebKit::RemoteImageBuffer::create):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::didCreateImageBufferBackend):
(WebKit::RemoteRenderingBackend::shouldBlockIOSurfaceCreationForNextImageBufferForTesting):
(WebKit::RemoteRenderingBackend::createImageBufferWithQualifiedIdentifier):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.messages.in:
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::shouldBlockIOSurfaceCreationForNextImageBufferForTesting):
(WebKit::RemoteRenderingBackendProxy::didCreateImageBufferBackend):
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.messages.in:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::shouldBlockIOSurfaceCreationForNextImageBuffer):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:

Canonical link: <a href="https://commits.webkit.org/252493@main">https://commits.webkit.org/252493@main</a>
</pre>

smfr commented on Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.messages.in:
```diff
@@ -23,7 +23,7 @@
 
 messages -> RemoteRenderingBackendProxy NotRefCounted {
-    DidCreateImageBufferBackend(WebKit::ImageBufferBackendHandle handle, WebCore::RenderingResourceIdentifier renderingResourceIdentifier)
+    DidCreateImageBufferBackend(WebKit::ImageBufferBackendHandle handle, WebCore::RenderingResourceIdentifier renderingResourceIdentifier, bool didFailToCreateIOSurface)
```
I think you should send a RenderingMode back (like `RenderingMode renderingModeOfAllocatedBuffer`).

nmoucht replied:
Fixed.

shallawa commented on Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
```diff
@@ -390,12 +397,15 @@ void RemoteRenderingBackendProxy::finalizeRenderingUpdate()
-
+    // If we receive that we failed to create an IOSurface from the GPU process, fallback to an unaccelerated buffer
+    if (imageBuffer->renderingMode() == RenderingMode::Accelerated && didFailToCreateIOSurface)
+        imageBuffer->fallbackToUnaccelerated();
```
I think changing the renderingMode is not sufficient. I think you need to do something like this:

```
const auto& parameters = imageBuffer->parameters();
auto backendInfo = RemoteRenderingBackendProxy::populateBackendInfo<UnacceleratedImageBufferShareableBackend>(parameters);
imageBuffer->setBackendInfo(backendInfo);
```

The backendInfo includes other data which depends on the backend type like 'memoryCost'.

nmoucht replied:
Fixed.

shallawa commented on Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
```diff
@@ -390,12 +397,15 @@ void RemoteRenderingBackendProxy::finalizeRenderingUpdate()
 }
 
-void RemoteRenderingBackendProxy::didCreateImageBufferBackend(ImageBufferBackendHandle handle, RenderingResourceIdentifier renderingResourceIdentifier)
+void RemoteRenderingBackendProxy::didCreateImageBufferBackend(ImageBufferBackendHandle handle, RenderingResourceIdentifier renderingResourceIdentifier, bool didFailToCreateIOSurface)
```
Why do we need this flag 'didFailToCreateIOSurface'? Can't we use something like this:

`bool didFailToCreateAcceleratedBackend = imageBuffer->renderingMode() == RenderingMode::Accelerated && std::holds_alternative<ShareableBitmap::Handle>(handle);`

nmoucht replied:
I changed it to send back the rendering mode used by the gpu process. I think that makes it a bit clearer than using the handle.

nmoucht replied:
After further discussion, I will go with this approach so that we will not have redundant parameters.

shallawa commented on Source/WebCore/page/ChromeClient.h:
```diff
@@ -364,6 +364,7 @@ class ChromeClient {
     virtual DisplayRefreshMonitorFactory* displayRefreshMonitorFactory() const { return nullptr; }
 
     virtual RefPtr<ImageBuffer> createImageBuffer(const FloatSize&, RenderingMode, RenderingPurpose, float, const DestinationColorSpace&, PixelFormat) const { return nullptr; }
+    virtual void shouldBlockIOSurfaceCreationForNextImageBuffer() { }
```
Instead of going through sending this message from WebProcess to GPUProcess, can't we do the following:

Skip this code in ImageBufferIOSurfaceBackend.cpp if the Internals function is called.

```
    IntSize maxSize = IOSurface::maximumSize();
    if (backendSize.width() > maxSize.width() || backendSize.height() > maxSize.height())
        return { };

```

In the test, create the largest backend for a canvas  `<canvas width="4096" height="4096">`

So we will be trying to create an IOSurface which is larger than IOSurface::maximumSize() in WebProcess. But the fallback will happen in the GPUProcess.

nmoucht replied:
Fixed.

cdumez commented on Source/WebCore/platform/graphics/ImageBuffer.h:
```diff
@@ -155,6 +155,10 @@ class ImageBuffer : public ThreadSafeRefCounted<ImageBuffer, WTF::DestructionThr
     AffineTransform baseTransform() const { return m_backendInfo.baseTransform; }
     size_t memoryCost() const { return m_backendInfo.memoryCost; }
     size_t externalMemoryCost() const { return m_backendInfo.externalMemoryCost; }
+    void setBackendInfo(ImageBufferBackend::Info& parameters)
```
Should either be a `const ImageBufferBackend::Info&` or `ImageBufferBackend::Info&&`. Here, I suggest `ImageBufferBackend::Info&&`

nmoucht replied:
I agree. Changed to `ImageBufferBackend::Info&&`.

cdumez commented on Source/WebCore/platform/graphics/ImageBuffer.h:
```diff
@@ -155,6 +155,10 @@ class ImageBuffer : public ThreadSafeRefCounted<ImageBuffer, WTF::DestructionThr
     size_t externalMemoryCost() const { return m_backendInfo.externalMemoryCost; }
+    void setBackendInfo(ImageBufferBackend::Info& parameters)
+    {
+        m_backendInfo = parameters;
```
`m_backendInfo = WTFMove(parameters);`

nmoucht replied:
Fixed.

cdumez commented on Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
```diff
@@ -390,18 +395,24 @@ void RemoteRenderingBackendProxy::finalizeRenderingUpdate()
         imageBuffer->setBackend(AcceleratedImageBufferRemoteBackend::create(imageBuffer->parameters(), WTFMove(handle)));
+        backendInfo = ImageBuffer::populateBackendInfo<AcceleratedImageBufferRemoteBackend>(imageBuffer->parameters());
+    }
+    imageBuffer->setBackendInfo(backendInfo);
```
`imageBuffer->setBackendInfo(WTFMove(backendInfo));`

nmoucht replied:
Fixed.

cdumez commented on Source/WebCore/testing/Internals.cpp:
```diff
@@ -6827,4 +6827,19 @@ String Internals::modelInlinePreviewUUIDForModelElement(const HTMLModelElement&
 
 #endif
 
+void Internals::shouldBlockIOSurfaceCreationForNextImageBuffer()
```
This is not a good name, it looks like a getter but doesn't return anything. If a function doesn't return a bool, it should probably not start with "should" (or "is" or similar prefixes).

nmoucht replied:
Changed to `blockIOSurfaceCreationForNextImageBuffer`.

cdumez commented on Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
```diff
@@ -185,6 +186,7 @@ class RemoteRenderingBackendProxy
 
     RenderingUpdateID m_renderingUpdateID;
     RenderingUpdateID m_didRenderingUpdateID;
+    bool m_shouldBlockIOSurfaceCreationForNextImageBuffer;
```
This is missing a default initializer so it is undefined what its initial value will be.

nmoucht replied:
Fixed.

smfr commented on Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
```diff
@@ -137,13 +137,13 @@ RefPtr<ImageBuffer> RemoteRenderingBackendProxy::createImageBuffer(const FloatSi
         // Web Content process, so they can be painted into the tiles.
         if (!WebProcess::singleton().shouldUseRemoteRenderingFor(RenderingPurpose::DOM))
-            imageBuffer = RemoteImageBufferProxy::create<AcceleratedImageBufferShareableMappedBackend>(size, resolutionScale, colorSpace, pixelFormat, purpose, *this);
+            imageBuffer = RemoteImageBufferProxy::create<AcceleratedImageBufferShareableMappedBackend>(size, resolutionScale, colorSpace, pixelFormat, purpose, *this, std::exchange(m_blockIOSurfaceCreationForNextImageBuffer, false));
```
We avoid `false` parameters (read up on the "boolean trap"). Named enums are better.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/testing/Internals.cpp:
```diff
@@ -6827,4 +6827,19 @@ String Internals::modelInlinePreviewUUIDForModelElement(const HTMLModelElement&
+    auto* page = document->page();
+    if (!page)
+        return;
+    page->settings().setMaximumAccelerated2dCanvasSize(UINT_MAX);
```
I don't understand the need to mess with these size limits.

nmoucht replied:
Renamed to avoidIOSurfaceSizeCheckInWebProcess to make this more clear.

smfr commented on Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
```diff
@@ -924,6 +924,12 @@ RefPtr<ImageBuffer> WebChromeClient::createImageBuffer(const FloatSize& size, Re
+
+void WebChromeClient::blockIOSurfaceCreationForNextImageBuffer()
+{
+    m_page.ensureRemoteRenderingBackendProxy().blockIOSurfaceCreationForNextImageBufferForTesting();
```
This needs some kind of `WebProcess::singleton().shouldUseRemoteRenderingFor(purpose)` check so that the test gives sensible behavior when "Canvas in GPU process" is disabled.

nmoucht replied:
Fixed.

shallawa commented on Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.h:
```diff
@@ -44,10 +44,10 @@ class RemoteRenderingBackend;
     {
         auto parameters = WebCore::ImageBufferBackend::Parameters { size, resolutionScale, colorSpace, pixelFormat, purpose };
-        if (BackendType::calculateSafeBackendSize(parameters).isEmpty())
+        if (BackendType::calculateSafeBackendSize(parameters).isEmpty() && avoidIOSurfaceSizeCheckInWebProcess == RemoteRenderingBackendProxy::AvoidIOSurfaceSizeCheckInWebProcess::No)
```
I think the order should be reversed:
`
if (avoidIOSurfaceSizeCheckInWebProcess == RemoteRenderingBackendProxy::AvoidIOSurfaceSizeCheckInWebProcess::No && BackendType::calculateSafeBackendSize(parameters).isEmpty())
`

nmoucht replied:
Fixed.

shallawa commented on Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
```diff
@@ -153,6 +153,11 @@ RefPtr<ImageBuffer> RemoteRenderingBackendProxy::createImageBuffer(const FloatSi
     return nullptr;
 }
 
+void RemoteRenderingBackendProxy::avoidIOSurfaceSizeCheckInWebProcessForTesting()
```
I think this should be named "setAvoid...()" since its sets a flag called "m_avoid...". It can also be inlined in the header file.

nmoucht replied:
This is now on HTMLCanvasElement, but I changed the name.

shallawa commented on Source/WebCore/platform/graphics/ImageBuffer.h:
```diff
@@ -155,6 +155,10 @@ class ImageBuffer : public ThreadSafeRefCounted<ImageBuffer, WTF::DestructionThr
+    void setBackendInfo(ImageBufferBackend::Info&& parameters)
+    {
+        m_backendInfo = WTFMove(parameters);
+    }
```
I think either this function is written in one line or moved to the source file.

nmoucht replied:
Fixed.

shallawa commented on Source/WebCore/testing/Internals.cpp:
```diff
@@ -6827,4 +6827,19 @@ String Internals::modelInlinePreviewUUIDForModelElement(const HTMLModelElement&
+    page->settings().setMaximumAccelerated2dCanvasSize(UINT_MAX);
+    HTMLCanvasElement::setMaxCanvasAreaForTesting(UINT_MAX);
+    HTMLCanvasElement::setMaxPixelMemoryForTesting(UINT_MAX);
+    page->chrome().client().avoidIOSurfaceSizeCheckInWebProcess();
```
It is weird that we set 'avoidIOSurfaceSizeCheckInWebProcess' in RemoteRenderingBackendProxy via ChromeClient but use it only once for the first RemoteImageBufferProxy we create. Can't we apply it to the canvas element and propagate it from the HTMLCanvasElement to its RemoteImageBufferProxy::create()?
`
void Internals::avoidIOSurfaceSizeCheckInWebProcess(HTMLCanvasElement& element)
`

nmoucht replied:
Fixed.

shallawa commented on LayoutTests/fast/canvas/image-buffer-iosurface-disabled.html:
```diff
@@ -0,0 +1,15 @@
+<div id=div>text</div>
```
This element can contain the message below:
```
<div>This test passes if it doesn't crash</div>
```

nmoucht replied:
Fixed.

shallawa commented on LayoutTests/fast/canvas/image-buffer-iosurface-disabled.html:
```diff
@@ -0,0 +1,15 @@
+if (window.testRunner)
+    testRunner.dumpAsText();
+
+div.offsetLeft;
```
I do not think you need this.

nmoucht replied:
Fixed.

shallawa commented on LayoutTests/fast/canvas/image-buffer-iosurface-disabled.html:
```diff
@@ -0,0 +1,15 @@
+    internals.avoidIOSurfaceSizeCheckInWebProcess(canvas);
+var context = canvas.getContext("2d");
+context.fillRect(0,0,canvas.width, canvas.height);
+div.textContent = "This test passes if it doesn't crash";
```
No need to set the div element textContent if getContext() passes. You can set it above.

nmoucht replied:
Fixed.

shallawa commented on Source/WebCore/html/HTMLCanvasElement.h:
```diff
@@ -186,7 +187,8 @@ class HTMLCanvasElement final : public HTMLElement, public CanvasBase, public Ac
     std::optional<bool> m_usesDisplayListDrawing;
-
+    
+    bool m_avoidIOSurfaceSizeCheckInWebProcess { false };
```
I think this member should include "ForTesting" so its use in the code becomes clearer. And I am not sure why we need to include "InWebProcess" or "IOSurface". The name should be generic and should not be specific for a certain backend or where it is used.

Can't it be `m_avoidBackendSizeCheckForTesting` or just `m_avoidSizeCheckForTesting`?

nmoucht replied:
Fixed.

shallawa commented on Source/WebCore/platform/HostWindow.h:
```diff
@@ -66,7 +66,7 @@ class HostWindow {
     virtual IntRect rootViewToAccessibilityScreen(const IntRect&) const = 0;
 
-    virtual RefPtr<ImageBuffer> createImageBuffer(const FloatSize&, RenderingMode, RenderingPurpose, float resolutionScale, const DestinationColorSpace&, PixelFormat) const = 0;
+    virtual RefPtr<ImageBuffer> createImageBuffer(const FloatSize&, RenderingMode, RenderingPurpose, float resolutionScale, const DestinationColorSpace&, PixelFormat, bool) const = 0;
```
I think we should include the argument name and its default value as well to be consistent its subclass `Chrome`

```
..., bool avoidIOSurfaceSizeCheckInWebProcess = false);
```

nmoucht replied:
Fixed.

shallawa commented on Source/WebCore/html/HTMLCanvasElement.cpp:
```diff
@@ -926,7 +931,7 @@ void HTMLCanvasElement::createImageBuffer() const
     }();
 
-    setImageBuffer(ImageBuffer::create(size(), RenderingPurpose::Canvas, 1, colorSpace, pixelFormat, bufferOptions, { hostWindow }));
+    setImageBuffer(ImageBuffer::create(size(), RenderingPurpose::Canvas, 1, colorSpace, pixelFormat, bufferOptions, { hostWindow }, m_avoidIOSurfaceSizeCheckInWebProcess));
```
I think m_avoidIOSurfaceSizeCheckInWebProcess should be included in `ImageBuffer::CreationContext`

```
..., { hostWindow, m_avoidIOSurfaceSizeCheckInWebProcess });
```

I think this makes sense since hostWindow and m_avoidIOSurfaceSizeCheckInWebProcess control how the ImageBuffer is created.

nmoucht replied:
Fixed.

shallawa commented on Source/WebCore/platform/graphics/ImageBuffer.h:
```diff
@@ -69,7 +69,7 @@ class ImageBuffer : public ThreadSafeRefCounted<ImageBuffer, WTF::DestructionThr
     };
 
-    WEBCORE_EXPORT static RefPtr<ImageBuffer> create(const FloatSize&, RenderingPurpose, float resolutionScale, const DestinationColorSpace&, PixelFormat, OptionSet<ImageBufferOptions> = { }, const CreationContext& = { });
+    WEBCORE_EXPORT static RefPtr<ImageBuffer> create(const FloatSize&, RenderingPurpose, float resolutionScale, const DestinationColorSpace&, PixelFormat, OptionSet<ImageBufferOptions> = { }, const CreationContext& = { }, bool avoidIOSurfaceSizeCheckInWebProcess = false);
```
No need for this if avoidIOSurfaceSizeCheckInWebProcess is added to CreationContext.

nmoucht replied:
Fixed.

shallawa commented on Source/WebCore/platform/graphics/ImageBuffer.cpp:
```diff
@@ -51,7 +51,7 @@ namespace WebCore {
 static const float MaxClampedArea = MaxClampedLength * MaxClampedLength;
 
-RefPtr<ImageBuffer> ImageBuffer::create(const FloatSize& size, RenderingPurpose purpose, float resolutionScale, const DestinationColorSpace& colorSpace, PixelFormat pixelFormat, OptionSet<ImageBufferOptions> options, const CreationContext& creationContext)
+RefPtr<ImageBuffer> ImageBuffer::create(const FloatSize& size, RenderingPurpose purpose, float resolutionScale, const DestinationColorSpace& colorSpace, PixelFormat pixelFormat, OptionSet<ImageBufferOptions> options, const CreationContext& creationContext, bool avoidIOSurfaceSizeCheckInWebProcess)
```
No need for this if avoidIOSurfaceSizeCheckInWebProcess is added to CreationContext.

nmoucht replied:
Fixed.

shallawa commented on Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
```diff
@@ -922,8 +922,9 @@ RefPtr<ImageBuffer> WebChromeClient::createImageBuffer(const FloatSize& size, Re
-    return m_page.ensureRemoteRenderingBackendProxy().createImageBuffer(size, renderingMode, purpose, resolutionScale, colorSpace, pixelFormat);
+    return m_page.ensureRemoteRenderingBackendProxy().createImageBuffer(size, renderingMode, purpose, resolutionScale, colorSpace, pixelFormat, avoidIOSurfaceSizeCheckInWebProcess);
 }
+
```
This extra line is not needed.

nmoucht replied:
Fixed.

shallawa commented on Source/WebCore/testing/Internals.idl:
```diff
@@ -1159,4 +1159,5 @@ typedef (FetchRequest or FetchResponse) FetchObject;
 
     [Conditional=ARKIT_INLINE_PREVIEW_MAC] Promise<sequence<DOMString>> modelInlinePreviewUUIDs();
     [Conditional=ARKIT_INLINE_PREVIEW_MAC] DOMString modelInlinePreviewUUIDForModelElement(HTMLModelElement modelElement);
+    undefined avoidIOSurfaceSizeCheckInWebProcess(Element element);
```
Is there a reason this API is not passed an `HTMLCanvasElement`?

nmoucht replied:
This was originally due to a linker error. Adding ExportMacro=WEBCORE_EXPORT to HTMLCanvasElement fixed this issue.

shallawa commented on Source/WebCore/testing/Internals.cpp:
```diff
@@ -6827,4 +6827,20 @@ String Internals::modelInlinePreviewUUIDForModelElement(const HTMLModelElement&
+    HTMLCanvasElement::setMaxCanvasAreaForTesting(UINT_MAX);
+    HTMLCanvasElement::setMaxPixelMemoryForTesting(UINT_MAX);
+    if (is<HTMLCanvasElement>(element))
+        downcast<HTMLCanvasElement>(element).setAvoidIOSurfaceSizeCheckInWebProcessForTesting();
```
Since this API is not going to work correctly if the element is not `HTMLCanvasElement`, why do not we pass an `HTMLCanvasElement` instead of passing `Element`?

nmoucht replied:
Fixed.

shallawa commented on Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
```diff
@@ -395,7 +395,12 @@ void RemoteRenderingBackendProxy::didCreateImageBufferBackend(ImageBufferBackend
-
+    
+    // If web process tried to create an accelerated buffer but gets that it created an unaccelerated buffer from the GPU process
+    // fallback to unaccelerated buffer in the web process
```
I think this comment needs word smithing and I am not sure if it adds much value than describing the code below it.

nmoucht replied:
I removed this comment.

nmoucht committed:
Handle failed IOSurface creation for image buffer
https://bugs.webkit.org/show_bug.cgi?id=242556
<rdar://94622888>

Reviewed by Said Abou-Hallawa.

Check the backend handle sent back to the web process to ensure the image buffer in the web process
is in synch with the GPU process. For the test, create a canvas greater than the max IOSurface size,
disabling the check in the web process so that an accelerated buffer is created, IOSurface
creation in the GPU process fails so it creates an unaccelerated buffer, and ensure that the
web process correctly changes the backend type to an unaccelerated buffer.

* LayoutTests/fast/canvas/image-buffer-iosurface-disabled-expected.txt: Added.
* LayoutTests/fast/canvas/image-buffer-iosurface-disabled.html: Added.
* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::shouldBlockIOSurfaceCreationForNextImageBuffer):
* Source/WebCore/platform/graphics/ConcreteImageBuffer.h:
* Source/WebCore/platform/graphics/ImageBuffer.h:
(WebCore::ImageBuffer::fallbackToUnaccelerated):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::shouldBlockIOSurfaceCreationForNextImageBuffer):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/GPUProcess/graphics/RemoteImageBuffer.h:
(WebKit::RemoteImageBuffer::create):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::didCreateImageBufferBackend):
(WebKit::RemoteRenderingBackend::shouldBlockIOSurfaceCreationForNextImageBufferForTesting):
(WebKit::RemoteRenderingBackend::createImageBufferWithQualifiedIdentifier):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.messages.in:
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::shouldBlockIOSurfaceCreationForNextImageBufferForTesting):
(WebKit::RemoteRenderingBackendProxy::didCreateImageBufferBackend):
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.messages.in:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::shouldBlockIOSurfaceCreationForNextImageBuffer):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:

Canonical link: https://commits.webkit.org/252493@main
